### DISCUSSION TITLE
Form does not work

### DISCUSSION THREAD

**Discussion by aaaleeenaaa** [neuefische/questions#584] [Web] [answered]
Hello everyone, 

I have a form on a separate page where one thing is entered. Submitting the form should redirect to the home page where each entered location then is rendered within a list as a card. Actually probably not that hard, but I thought about it so much and edited/deleted my code so many times that I'm now fully lost. But well, maybe having this in the discussion board might help others as well. 

This is the [US](https://github.com/aaaleeenaaa/capstone-project-bee-organized/issues/2) I am working on. 

I guess the problem is lifting states/passing props, but I am way too confused now. 

Currently, submitting the form leads to the index page but nothing is shown there. 

Some elements have some basic styling which I will not show here. That is why the names might be different than normally. 

Thanks in advance, appreciate even the smallest hints!

So that is my form: 
```
export default function Form({ location }) {
  const router = useRouter();
  // const { trigger } = useSWRMutation("/api/locations", sendRequest);

  function handleSubmit(event) {
    event.preventDefault();
    const formData = new FormData(event.target);
    const locationData = Object.fromEntries(formData);
    router.push("/");
  }

  return (
    <FormContainer onSubmit={handleSubmit} location={location}>
      <Label htmlFor="location">Add a new location</Label>
      <Input id="location" name="location" type="text" />
      <StyledSubmitButton type="submit">Submit</StyledSubmitButton>
    </FormContainer>
  );
}

```

It is supposed to be rendered on this page: 
```
export default function AddLocationPage() {
  const [location, setLocation] = useState("");
  const router = useRouter();

  async function addLocation(event) {
    const formData = new FormData(event.target);
    const placesData = Object.fromEntries(formData);
    trigger(placesData);
    router.push("/");
  }

  return <Form onSubmit={addLocation} />;
}
```

That is my card component:

```
export default function Card({ locationData }) {
  return (
    <StyledListElement>
      <StyledLink href="#">
        <CardElement h>
          <p>{locationData}</p>
        </CardElement>
      </StyledLink>
    </StyledListElement>
  );
}
```

That is the list component where I try to map over the data I retrieve from the form so that one card is rendered for each entered location:
```
export default function LocationList({ locationData }) {
  const router = useRouter();
  const { data, isLoading } = useSWR("/api/locations");

  if (!data) return;

  if (isLoading) {
    return <h1>Loading...</h1>;
  }

  return (
    <StyledList>
      {data.map((locationData) => (
        <Card location={locationData} key={locationData.index} />
      ))}
    </StyledList>
  );
}
```

And in the end, it should all be rendered on the index page. (AddButton is used to get to the page with the form, this works perfectly fine)

```
export default function HomePage() {
  const { data } = useSWR("/api/locations", { fallbackData: [] });

  return (
    <>
      <LocationList location={data} />

      <AddButton />
    </>
  );
}
```

**Comment by ThomasKirst** [comment]
Hi Alena,

This looks quite good so far. There is mainly one issue with your form / addLocation page:

```jsx
export default function AddLocationPage() {
  const [location, setLocation] = useState("");
  const router = useRouter();

  async function handleAddLocation(event) {
    const formData = new FormData(event.target);
    const placesData = Object.fromEntries(formData);
    trigger(placesData);
    router.push("/");
  }

  // Your page passes the function `onAddLocation` down to the form. The form expects a prop called `onSubmit`.
  // But your form just "listens" for a `location` prop.
  // I would rename `onSubmit` to `onAddLocation` and this functions needs to be called afterwards in your form.
  return <Form onAddLocation={handleAddLocation} />;
}
```

```jsx
//export default function Form({ location }) {
export default function Form({ onAddLocation }) {
  const router = useRouter();

  function handleSubmit(event) {
    event.preventDefault();
    const formData = new FormData(event.target);
    const locationData = Object.fromEntries(formData);
    onAddLocation(locationData);
    router.push("/");
  }
 
  // 👇 You will not need the location attribute on the FormContainer here, I guess
  return (
    <FormContainer onSubmit={handleSubmit} location={location}>
      <Label htmlFor="location">Add a new location</Label>
      <Input id="location" name="location" type="text" />
      <StyledSubmitButton type="submit">Submit</StyledSubmitButton>
    </FormContainer>
  );
}

```

With these few adjustments you will have the state stored in your page `AddLocationPage`.
If you want to have that information even higher up in the component tree you would have to move it to your _app.js in order to share the locations with other pages.

Let me know if this works for you :)
– Thomas

**Reply by ThomasKirst** [reply]
I found another small improvement for your `addLocation` function.

```jsx
export default function AddLocationPage() {
  const [location, setLocation] = useState("");
  const router = useRouter();

  // The location will be passed here from the form (when `onAddLocation` is called)
  // The function does not need to be async
 function handleAddLocation(newLocation) {
    setLocation(newLocation)
    /*
    const formData = new FormData(event.target);
    const placesData = Object.fromEntries(formData);
    trigger(placesData);
    router.push("/");
    */
  }

  return <Form onAddLocation={handleAddLocation} />;
}
```

**Comment by aaaleeenaaa** [comment]
Hi Thomas, 

thanks a lot for your help. I changed my code according to your suggestions.

However, the result is the same as before. When I submit the form, I am redirected to the index page but nothing is shown. 

I need the information about the locations not on the AddLocationPage but in the LocationList component that should be rendered on the index page. As the index page is just another page, I believe that it is on the same level in the component tree as the AddLocationPage and the LocationList (as it is just a component and not a page) is a level below? (Or is that wrong? I am totally confused by now)

Of course, I also tried to put 

``
 const [location, setLocation] = useState("");
``

in the _app.js. Then, I get an error message that the "setLocation" within the handleAddLocation is not defined. However, moving that to the _app.js did not make any difference as well. 

Maybe you could have another look at it? Thanks in advance!

**Reply by NikRoe** [reply]
Hi Alena,

can you send us a PR of the current branch? 

Thanks in advance :)

**Comment by aaaleeenaaa** [comment]
Hi Niklas, 

[Here is the link to the pull request of the current status](https://github.com/aaaleeenaaa/capstone-project-bee-organized/pull/9)! Probably kind of a mess, I really lost the overview by now. Thanks a lot!

**Reply by NikRoe** [reply, marked as answer]
Hi there,

we created a fork and a PR where you can see the needed changes for your code to run again:
https://github.com/aaaleeenaaa/capstone-project-bee-organized/pull/10/files

Please do not merge the PR but implement those changes into your project manually on your own.
We don't want our names, or in this case mine, to appear as a contributor in your capstone-project.
Once you have implemented everything and it works, you can close the PR. 

Hope this helps =) 

Edit: 
What we did to make the app work again: 
We moved the state and the handler function to the _app.js and can pass the handler function as well as the state as props to the pages from the _app.js.

The AddLocationPage accepts the handler function as a prop, calls the Form component and passes that function as a prop one level down.
The Form component calls the now called onAddLocation within the handleSubmit()

The HomePage receives the locations as a prop, calls the LocationList component and passes the locations as a prop one level down
The LocationList component receives the locations as a props and maps over the array to display all locations.
Since there is no id within the location object, we currently use the index, which should be replaced with an id as soon as possible.

**Reply by aaaleeenaaa** [reply]
It works! Thank you so so so so much! As suggested, I will close the PR now.
